hibernate-core/src/main/java/org/hibernate/internal/util/collections/InstanceIdentityMap.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.internal.util.collections;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.hibernate.engine.spi.InstanceIdentity;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
+
+/**
+ * Utility collection backed by {@link PagedArray} that takes advantage of {@link InstanceIdentity}'s
+ * unique identifier to store objects and provide {@link Map}-like functionalities.
+ * <p>
+ * Methods accessing / modifying the map with {@link Object} typed parameters will need
+ * to type check against the instance identity interface which might be inefficient,
+ * so it's recommended to use the position (int) based variant of those methods.
+ * <p>
+ * Iterating through the whole map is most efficient with {@link #forEach}, and since
+ * we simply iterate the underlying array, it's also concurrent and reentrant safe.
+ */
+public class InstanceIdentityMap<K extends InstanceIdentity, V> implements Map<K, V> {
```

Comment:
I know that is tempting to implements Map interface but I would create a bespoke API based on what the usage you expect to have - which means that "maybe" ( didn't checked actually) there's no point for returning the keys or just expose anything "key related"

Reply:
The problem I see is due to the increased allocation rate due to additional entries.
I will still keep the map case separate day the other one where implementing Map is not needed

Reply:
We have the problem of instances coming from one session to the other with their instance-id already set, and to ensure they are the same we would need an `==` check on the key stored in the entry.

Also, we have operations that iterate the map and act not only on its values but also its keys, so I believe creating the entries is not avoidable even if we were to not implement `Map` on the new collection :(

Reply:
We not have 2 separate collections: a bespoke `InstanceIdentityStore` API, which is used for entities and allows us to keep allocations under control, and a `InstanceIdentityMap` which implements `java.util.Map` and is used for collections where we often need to iterate and act on the keys.